File: LeetCode/Medium/0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int count = 0;

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == '1') {
                    dfs(i, j, grid);
                    count++;
                }
            }
        }

        return count;
    }

    private void dfs(int x, int y, char[][] grid) {
        if (x < 0 || x >= grid.length || y < 0 || y >= grid[0].length || grid[x][y] == '0') return;

        grid[x][y] = '0';
        dfs(x, y + 1, grid);
        dfs(x, y - 1, grid);
        dfs(x - 1, y, grid);
        dfs(x + 1, y, grid);
    }
}

File: LeetCode/Medium/0200-number-of-islands/0200-number-of-islands.kt
class Solution {
    fun numIslands(grid: Array<CharArray>): Int {
        fun dfs(x: Int, y: Int) {
            when {
                (x < 0) ||
                (x >= grid.size) ||
                (y < 0) ||
                (y >= grid[0].size) ||
                (grid[x][y] == '0') -> return
            }

            grid[x][y] = '0'
            dfs(x, y + 1)
            dfs(x, y - 1)
            dfs(x - 1, y)
            dfs(x + 1, y)
        }
        
        var count = 0

        for (i in grid.indices) {
            for (j in grid[0].indices) {
                if (grid[i][j] == '1') {
                    dfs(i, j)
                    count++
                }
            }
        }

        return count
    }
}